### PR TITLE
[distributed] cleanup dist autograd context on other nodes when it is released on one node

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#27951 [distributed] cleanup dist autograd context on other nodes when it is released on one node**

Per https://github.com/pytorch/pytorch/issues/25525, we want to clean up the distributed autograd context across the other nodes when a single node is done (here done means exited the context manager `with dist_autograd.context() as context_id: ...`).

This PR does a few things to implement the above:
1) Add classes to encapsulate messages for requesting this context release and the response
2) Handling of this request in `request_callback_impl.cpp`. When we receive this request, we get the context from a given context_id and release it.
3) RPC call in `DistAutogradContainer::releaseContext` to send this command. This currently does not wait for an ack or implement any sort of retrying. We send the RPC to all the workerIds we have come into contact with (implemented in https://github.com/pytorch/pytorch/pull/26324)
4) Relevant unit tests

In follow up PRs, we will add error checking + retries for this call.
Differential Revision: [D17920137](https://our.internmc.facebook.com/intern/diff/D17920137/)

### CODE REVIEW

**Comment by rohan-varma** [comment]
@pritamdamania87 I think this is ready for a first pass look now. some legit failures (https://circleci.com/gh/pytorch/pytorch/3209870?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link) in dist autograd tests since the contexts are being cleaned up across nodes now, so the tests might need some more refactoring.

File: torch/csrc/distributed/rpc/request_callback_impl.cpp
Comment (rohan-varma) [inline torch/csrc/distributed/rpc/request_callback_impl.cpp]: Actually, we probably need to do this check + release atomically, since it is still possible that another thread could release the context after we've done the `hasContextWithId` check. Then, `releaseContext` would throw but we don't necessarily want to throw, since the context has been freed as desired. Should we make a version of releaseContext that does not throw if the context we are releasing does not exist (or control this with a param?)

File: test/dist_autograd_test.py
Comment (pritamdamania87) [inline test/dist_autograd_test.py]: Add a small comment here mentioning we don't need a lock here since we hold the GIL when executing python UDFs.

File: test/dist_autograd_test.py
Comment (pritamdamania87) [inline test/dist_autograd_test.py]: Lets use `dist.barrier()` since `sync_rpc` might go away in a few future changes we have planned.

File: test/dist_autograd_test.py
Comment (pritamdamania87) [inline test/dist_autograd_test.py]: dist.barrier().

File: torch/csrc/distributed/autograd/context/dist_autograd_container.cpp
Comment (pritamdamania87) [inline torch/csrc/distributed/autograd/context/dist_autograd_container.cpp]: nit: Try to order includes alphabetically. I thought clang-tidy would catch this? Maybe it doesn't?

File: torch/csrc/distributed/autograd/context/dist_autograd_container.cpp
Comment (pritamdamania87) [inline torch/csrc/distributed/autograd/context/dist_autograd_container.cpp]: nit: `hasContextId`

File: torch/csrc/distributed/rpc/request_callback_impl.cpp
Comment (pritamdamania87) [reply torch/csrc/distributed/rpc/request_callback_impl.cpp]: Yes, let's do check + release automatically and have a version which doesn't throw if the context is cleaned up (can call it releaseContextIfPresent).

Comment (pritamdamania87) [reply torch/csrc/distributed/rpc/request_callback_impl.cpp]: Also, shouldn't `notifyWorkers` be true here? In case we do nested RPCs, we need this node to notify other nodes too.

File: torch/csrc/distributed/autograd/context/dist_autograd_container.cpp
Comment (pritamdamania87) [inline torch/csrc/distributed/autograd/context/dist_autograd_container.cpp]: In which case would `notifyWorkers` be false?

File: torch/csrc/distributed/autograd/context/dist_autograd_container.cpp
Comment (pritamdamania87) [inline torch/csrc/distributed/autograd/context/dist_autograd_container.cpp]: Create a gh issue mentioning that we should store the futures we get from these calls in a structure and retry them later if they failed.

File: test/dist_autograd_test.py
Comment (pritamdamania87) [inline test/dist_autograd_test.py]: Create a gh issue that we should add a test with multiple nested RPCs (with Python UDFs) once https://github.com/pytorch/pytorch/pull/27576 lands.

File: torch/csrc/distributed/autograd/rpc_messages/cleanup_autograd_context_req.cpp
Comment (pritamdamania87) [inline torch/csrc/distributed/autograd/rpc_messages/cleanup_autograd_context_req.cpp]: nit: `std::vector<at::IValue> ivalues {context_id_};`

File: torch/csrc/distributed/autograd/rpc_messages/cleanup_autograd_context_req.cpp
Comment (pritamdamania87) [inline torch/csrc/distributed/autograd/rpc_messages/cleanup_autograd_context_req.cpp]: Why do we need to create a tuple here? We just have one IValue, can't we just pickle that?

File: torch/csrc/distributed/autograd/rpc_messages/cleanup_autograd_context_resp.h
Comment (pritamdamania87) [inline torch/csrc/distributed/autograd/rpc_messages/cleanup_autograd_context_resp.h]: Mention that this is an empty response for the `CleanupAutogradContextReq`. The current comment seems to indicate this call is responsible for cleaning up the context.

**Review by pritamdamania87 — CHANGES_REQUESTED** [review]

File: torch/csrc/distributed/autograd/context/dist_autograd_container.cpp
Comment (rohan-varma) [reply torch/csrc/distributed/autograd/context/dist_autograd_container.cpp]: This is strange, when I run the auto linter it gives me this ordering. If I change the include order there is a linter error.

File: torch/csrc/distributed/autograd/context/dist_autograd_container.cpp
Comment (rohan-varma) [reply torch/csrc/distributed/autograd/context/dist_autograd_container.cpp]: Right now, this is here since this PR doesn't support cycle detection, so this is a simple flag to avoid cycles. In a follow up PR I was planning to add additional changes to supported nested RPCs. 

More generally, there may be cases where we know that we don't want to identify additional workers (for example if we know all of them have already cleaned up the context), then we can avoid extra RPCs with this flag.

File: torch/csrc/distributed/rpc/request_callback_impl.cpp
Comment (rohan-varma) [reply torch/csrc/distributed/rpc/request_callback_impl.cpp]: It should eventually be true, but since this PR doesn't implement cycle detection this is a way of not doing additional RPC calls. Once #27576 lands we can modify this to support nested rpc calls.

File: test/dist_autograd_test.py
Comment (rohan-varma) [reply test/dist_autograd_test.py]: I don't think `dist.barrier()` works here (just tried it and the tests fail if we replace it). It appears that `dist.barrier()` blocks until all the processes enter the barrier. But `rpc.sync_rpc()` gives us the further guarantee that all outstanding RPCs will be completed, which is what we need here, since we need the RPCs for releasing the context to finish. With `dist.barrier()` there is no guarantee that all outstanding RPCs will be completed, it looks like. cc @mrshenli in case my understanding isn't correct

File: torch/csrc/distributed/autograd/rpc_messages/cleanup_autograd_context_req.cpp
Comment (rohan-varma) [reply torch/csrc/distributed/autograd/rpc_messages/cleanup_autograd_context_req.cpp]: I was thinking of keeping it this way since we could potentially avoid a refactor if we decide to start sending additional data for this RPC.

File: torch/csrc/distributed/rpc/request_callback_impl.cpp
Comment (pritamdamania87) [reply torch/csrc/distributed/rpc/request_callback_impl.cpp]: Do we need to do cycle detection? If we do an atomic check + release, we might ask a node to release again, but that would just be a noop right?

File: test/dist_autograd_test.py
Comment (pritamdamania87) [reply test/dist_autograd_test.py]: Hmm what is the reason `dist.barrier` doesn't work here? We are only doing sync RPCs in this test so all RPCs should have finished on all nodes when we are done with the barrier.

> Why is sync_rpc going away? Is there going to be any other way to enforce all processes finish all outstanding RPCs and then move forward (without joining)?

Because the user should wait on RPCs that they care about and not the framework. If you care about an RPC, you should use future.wait() to check its result.

File: torch/csrc/distributed/autograd/rpc_messages/cleanup_autograd_context_req.cpp
Comment (pritamdamania87) [reply torch/csrc/distributed/autograd/rpc_messages/cleanup_autograd_context_req.cpp]: For now, let's avoid using a tuple since it would have some overhead. If later on we need this we can change this accordingly.

File: test/dist_autograd_test.py
Comment (rohan-varma) [reply test/dist_autograd_test.py]: But the RPC that is sent to release the context on other nodes isn't synchronous right? It is sent with `agent->send()` which enqueues the send, and the user doesn't wait on this (i.e. when the context manager is exited, the RPCs are sent, but we don't wait on these RPCs). If we want, we can wait on the future inside `releaseContext`, but that would be a slowdown to the user.

File: torch/csrc/distributed/rpc/request_callback_impl.cpp
Comment (rohan-varma) [reply torch/csrc/distributed/rpc/request_callback_impl.cpp]: Yes, this is true. Though for performance reasons we'd probably want to avoid the extra RPC caused by asking a node to release again.

File: torch/csrc/distributed/autograd/context/dist_autograd_container.cpp
Comment (rohan-varma) [reply torch/csrc/distributed/autograd/context/dist_autograd_container.cpp]: This will actually go away since we are checking + releasing atomically with `releaseContextIfPresent`

File: test/dist_autograd_test.py
Comment (rohan-varma) [reply test/dist_autograd_test.py]: In this particular case `dist.barrier()` works since the async RPCs haven't been sent, but in the other test where we sync after exiting the context manager, we would need a way to make sure the RPCs are sent + processed before continuing with the test.

File: torch/csrc/distributed/autograd/rpc_messages/cleanup_autograd_context_req.cpp
Comment (rohan-varma) [reply torch/csrc/distributed/autograd/rpc_messages/cleanup_autograd_context_req.cpp]: No longer needed per your suggestion since we just have a single value we are pickling.

File: torch/csrc/distributed/autograd/context/dist_autograd_container.cpp
Comment (rohan-varma) [reply torch/csrc/distributed/autograd/context/dist_autograd_container.cpp]: Since we now have `releaseContextIfPresent`, `notifyWorkers` should no longer be needed since the RPC would just do a no-op if the context didn't exist, so we can get rid of `notifyWorkers`.

File: torch/csrc/distributed/autograd/context/dist_autograd_container.h
Comment (rohan-varma) [inline torch/csrc/distributed/autograd/context/dist_autograd_container.h]: These methods are added to reduce code duplication between `releaseContext` and `releaseContextIfPresent`.

File: test/dist_autograd_test.py
Comment (rohan-varma) [reply test/dist_autograd_test.py]: See above conversation regarding this

File: test/dist_autograd_test.py
Comment (rohan-varma) [reply test/dist_autograd_test.py]: > Because the user should wait on RPCs that they care about and not the framework.

Agreed. What about the case where our framework itself sends async RPCs that as part of handling a user request? This is basically what these cleanup context RPCs are. The user does not have any handle to the futures created by them, but we should at some point guarantee that they are done.

File: torch/csrc/distributed/rpc/request_callback_impl.cpp
Comment (pritamdamania87) [reply torch/csrc/distributed/rpc/request_callback_impl.cpp]: > Yes, this is true. Though for performance reasons we'd probably want to avoid the extra RPC caused by asking a node to release again.

True, although for now let's just go with a simple version and later on we can evaluate whether we need to optimize this further.

File: test/dist_autograd_test.py
Comment (pritamdamania87) [reply test/dist_autograd_test.py]: One way to fix this would be to have loop with a timeout which checks for the context and sleeps for a while if the context still exists. If we exit this loop and the timeout has expired, we fail the test. The timeout could be 10s to be conservative. An example:

```
start = time.time()
bool success = False
while time.time() - start < 10:
  if not dist_autograd.has_context_id(context_id):
    success = True
    break
self.assertTrue(success)
```

This structure can also be moved into a helper function that we can reuse in other tests.

File: torch/csrc/distributed/rpc/request_callback_impl.cpp
Comment (pritamdamania87) [inline torch/csrc/distributed/rpc/request_callback_impl.cpp]: nit: s/created/create

**Review by pritamdamania87 — APPROVED** [review]

File: test/dist_autograd_test.py
Comment (zhaojuanmao) [inline test/dist_autograd_test.py]: it can be deleted since it is not used anymore

File: torch/csrc/distributed/autograd/context/dist_autograd_container.h
Comment (zhaojuanmao) [inline torch/csrc/distributed/autograd/context/dist_autograd_container.h]: wondering why they can not be merged?

Comment (rohan-varma) [reply torch/csrc/distributed/autograd/context/dist_autograd_container.h]: One checks if the context exists, one throws if it does not. We can merge them if we have a flag `throwIfNotExists` or something similar, should we do this? cc @pritamdamania87

File: test/dist_autograd_test.py
Comment (rohan-varma) [inline test/dist_autograd_test.py]: @pritamdamania87 I implemented your suggestion and removed the use of `rpc.sync_rpc()` - is this 10s timeout ok or will we need a longer one for TSAN/ASAN tests?

Comment (pritamdamania87) [reply test/dist_autograd_test.py]: 10s should be sufficient.

File: torch/csrc/distributed/autograd/context/dist_autograd_container.h
Comment (pritamdamania87) [inline torch/csrc/distributed/autograd/context/dist_autograd_container.h]: Lets mention that this throws an exception if context_id not found.

File: torch/csrc/distributed/autograd/context/dist_autograd_container.h
Comment (pritamdamania87) [reply torch/csrc/distributed/autograd/context/dist_autograd_container.h]: Two separate methods seems a bit cleaner to me than having a flag.

**Comment by rohan-varma** [comment]
Waiting on #28312 to land before landing this, since that is more urgent and it would be delayed by required rebase/test fixes if this went in first.

File: test/dist_autograd_test.py
Comment (rohan-varma) [inline test/dist_autograd_test.py]: @zhaojuanmao with these changes, had to add `dist.barrier()`s here and in other tests so there is no races for the autograd context.

File: torch/csrc/distributed/autograd/context/dist_autograd_container.cpp
Comment (pietern) [reply torch/csrc/distributed/autograd/context/dist_autograd_container.cpp]: There must be a newline between the top include (for the header corresponding to the source file) and the other includes.

Comment (rohan-varma) [reply torch/csrc/distributed/autograd/context/dist_autograd_container.cpp]: I'll clean this up (and the other files that this is missing) as part of a later diff. Should we add a lint rule for this?

Comment (pietern) [reply torch/csrc/distributed/autograd/context/dist_autograd_container.cpp]: It's a clang-format rule, not a clang-tidy rule. We don't lint on clang-format yet (see #25700).

**Pull request closed without merging**
